cocos/core/scene-graph/node.ts

```diff
@@ -625,16 +626,30 @@ export class Node extends BaseNode {
         }
         let child: this; let dirtyBits = 0;
 
+        let childLPos: Vec3;
+        let childMat: Mat4;
+        let childPos: Vec3;
+        let curMat: Mat4;
         while (i) {
             child = array_a[--i];
             dirtyBits |= child._dirtyFlags;
             if (cur) {
                 if (dirtyBits & TransformBit.POSITION) {
-                    Vec3.transformMat4(child._pos, child._lpos, cur._mat);
-                    child._mat.m12 = child._pos.x;
-                    child._mat.m13 = child._pos.y;
-                    child._mat.m14 = child._pos.z;
-                    NodePool.setVec3(child._poolHandle, NodeView.WORLD_POSITION, child._pos);
+                    // Vec3.transformMat4(child._pos, child._lpos, cur._mat);
+                    childLPos = child._lpos;
+                    childMat = child._mat;
+                    childPos = child._pos;
+                    curMat = cur._mat;
+
+                    childPos.x = curMat.m12 + childLPos.x;
+                    childPos.y = curMat.m13 + childLPos.y;
+                    childPos.z = curMat.m14 + childLPos.z;
```

Comment:
This completely breaks the computation for child positions when the parent has a non-identity rotation.